test/test_models.py

```diff
@@ -69,16 +91,255 @@ def check_script(self, model, name):
             msg = str(e) + str(tb)
         self.assertTrue(scriptable, msg)
 
-    def _test_classification_model(self, name, input_shape):
-        # passing num_class equal to a number other than 1000 helps in making the test
-        # more enforcing in nature
-        model = models.__dict__[name](num_classes=50)
-        self.check_script(model, name)
-        model.eval()
-        x = torch.rand(input_shape)
-        out = model(x)
-        self.assertEqual(out.shape[-1], 50)
+    def _check_scriptable(self, model, expected):
+        if expected is None:  # we don't check scriptability for all models
+            return
+
+        actual = True
+        msg = ''
+        try:
+            torch.jit.script(model)
+        except Exception as e:
+            tb = traceback.format_exc()
+            actual = False
+            msg = str(e) + str(tb)
+        self.assertEqual(actual, expected, msg)
+
+
+class ClassificationCoverageTester(TestCase):
+
+    # Find all models exposed by torchvision.models factory methods (with assumptions)
+    def get_available_classification_models(self):
+        # TODO add a registration mechanism to torchvision.models
+        return [k for k, v in models.__dict__.items() if callable(v) and k[0].lower() == k[0] and k[0] != "_"]
+
+    # Recursively gather test methods from all classification testers
+    def get_test_methods_for_class(self, klass):
+        all_methods = inspect.getmembers(klass, predicate=inspect.isfunction)
+        test_methods = set([method[0] for method in all_methods if method[0].startswith('test_')])
+        for child in klass.__subclasses__():
+            test_methods = test_methods.union(self.get_test_methods_for_class(child))
+        return test_methods
+
+    # Verify that all models exposed by torchvision.models factory methods
+    #    have corresponding test methods
+    # NOTE This does not include some of the extra tests (such as Resnet
+    #    dilation) and says nothing about the correctness of the test, nor
+    #    of the model. It just enforces a naming scheme on the tests, and
+    #    verifies that all models have a corresponding test name.
+    def test_classification_model_coverage(self):
```

Comment:
Why aren't the `torchvision.models` factory methods tested directly?

Reply:
The generated tests were implicit and unintentional, and most of the tests were identical, except for a few that were added/tweaked in decision branches. Besides the fact that I was enhancing the test with correctness checks for each model, these generated tests exhibit at least two anti-patterns:

1) Behavior changes that depend on type should be handled by subclassing, not by conditionals.
2) This creates a state where I can add a new model, and a (potentially) passing test gets generated without my ever looking at it. If the passing test is correct, great, I lucked out - but if it's not I have no warning. This is why tests should be explicit and intentional.

In the new code:

1) Test methods are grouped by the models they test. Making a change to ResNet? All the tests for all variants are in one obviously-named class, and you don't have to tease them out of an introspection loop. Differences like the special input shape needed by Inception are handled in the subclass devoted to Inception, not by polluting a conditional branch of omnibus code that tries to cover every variant of every class.
2) If you make a new model, and give it a factory method in `torchvision.models`, `test_classification_model_coverage()` fails the next time you run tests; thus, the dev is nudged to intentionally create a new test that covers the new model/factory method.

So we get more maintainable tests, and lower overhead to having to change tests for new/updated models, and an automatic canary test for new factory methods in `torchvision.models`.

Looking ever further forward, our model offerings are likely to continue to grow, and (I hope) we'll continue building out the test suite, both to cover new models and give better coverage on existing ones. Keeping an increasingly complex test suite for a growing body of code under test in a single class governed by decision branches will turn ugly. The new scheme should have more extensibility and a lower cost of change.

Reply:
Generally I think it's best to avoid huge refactorings before they're clearly needed. In this case i'm not sure the need has been demonstrated. While the situation you're describing of increasingly complex conditionals seems plausible, it isn't the case today, and I think the updated version here is a substantial increase to the complexity of the code. 
This adds:
     - caching of model inputs per dimensionality
     - introspection of test coverage
     - two separate pathways for testing scriptability
     - two separate pathways for adding model tests. 

I don't think the possibility of someone adding a new model and there being no testing for it is likely. If someone adds a new model, like memory efficient densenet, they're expected to add a new test for it. I don't think this is a difficult standard to uphold as a contributor or reviewer.

Reply:
This refactoring *was* needed in order to work with the tests in isolation. Those implicit, unintentional tests were a PITA to isolate and change, which limits their utility. Also, the correctness checks themselves are an example of the building complexity - they are sometimes slightly different for different models, e.g. checking that the memory efficient Densenet implementation showed the same output as the vanilla version. To your other points:

* I'm not married to the test input caching, and if there's a demonstrable issue that I'm not aware of (as opposed to merely a prospective one), I'm happy to change it.
* The introspection was already happening. I only included it so we wouldn't lose the sense of security people seem to get from the implicit/unintentional tests being generated over the factory methods.
* If by "two separate pathways" you mean the difference between classification and other model types, that's true. My brief only covered Classification models. I think the other model types should be broken out into their own files (as their testing needs will likely differ), and updated to dump the opaque generated tests.
* Yes, there is more code, but it's also modularized in a way that explicitly maps the tests to the code under test, removes conditionals in favor of OO, and makes it trivially obvious to find where to make a change or set a breakpoint for a particular model with a text search. "Less code" is not, in every situation, the best variable to optimize for.

Your last point - about test being enforced by process - is orthogonal to the point I made above. If someone adds a model, never looks at the (implicit, unintentional) tests, and the automated tests all show green, the likelihood is against either the dev or a reviewer going deeper. The generated tests leave that door open. The new tests do nothing if the dev codes their tests first, and gives a nudge if they don't.